lib/ja_serializer/dsl.ex

```diff
@@ -72,30 +72,43 @@ defmodule JaSerializer.DSL do
   end
 
   @doc false
-  defmacro __before_compile__(_env) do
+  defmacro __before_compile__(env) do
     quote do
       def __relations,  do: @relations
       def __location,   do: @location
       def __attributes, do: @attributes
+
+      unquote(define_inlined_attributes_map(env))
+    end
+  end
+
+  defp define_inlined_attributes_map(env) do
+    attributes = Module.get_attribute(env.module, :attributes)
+    conn = quote do: conn
+    struct = quote do: struct
+
+    # Construct ASL for map with keys from attributes calling the attribute fn
+    body = {:%{}, [],
+      Enum.map(attributes, fn k -> {k, {k, [], [struct, conn]}} end)
+    }
+
+    quote do
+      @compile {:inline, inlined_attributes_map: 2}
+      def inlined_attributes_map(unquote(struct), unquote(conn)), do: unquote(body)
```

Comment:
Here I would like to define directly `attributes`. But when I tried that, it seemed that definitions in `__before_compile__` get pinned in code before other modules, and effectively this function got overridden by the one in `JaSerializer.Serializer.attributes`. 

So I generate other one and then generate the final `attributes` in the `__using__` block as it was originally.

There is probably better way to do this :)